### PR TITLE
gh-101773: Optimize creation of Fractions in private methods

### PR DESCRIPTION
**Pull request by skirpichev**
- removed private ``_normalize`` argument of the Fraction constructor
- added _from_coprime_ints() helper to make fractions from a pair of coprime ints

* Issue: gh-101773

### CODE REVIEW

File: Lib/fractions.py
Comment (eendebakpt) [inline Lib/fractions.py:764]:
```suggestion
            raise ZeroDivisionError('Fraction(0, %s)' % db)
```

File: Lib/fractions.py
Comment (eendebakpt) [inline Lib/fractions.py:820]:
```suggestion
                    raise ZeroDivisionError('Fraction(0, %s)' %
```

File: Lib/fractions.py
Comment (eendebakpt) [inline Lib/fractions.py]: If Fraction is really immutable we could optimize here by returning self in case `a._numerator` is positive. But i am not sure whether the gain is large enough for the added code complexity.

File: Lib/fractions.py
Comment (skirpichev) [reply Lib/fractions.py:820]: I would like to reproduce the current exception, e.g.:
```
>>> fractions.Fraction(0)**-1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/sk/src/cpython/Lib/fractions.py", line 805, in __pow__
    return Fraction(a._denominator ** -power,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/sk/src/cpython/Lib/fractions.py", line 281, in __new__
    raise ZeroDivisionError('Fraction(%s, 0)' % numerator)
ZeroDivisionError: Fraction(1, 0)
```

File: Lib/fractions.py
Comment (skirpichev) [reply Lib/fractions.py:764]: Ditto here.

File: Lib/fractions.py
Comment (skirpichev) [reply Lib/fractions.py]: > If Fraction is really immutable 

"[Fraction](https://docs.python.org/3/library/fractions.html?highlight=fraction#fractions.Fraction) instances are hashable, and should be treated as immutable." (c) docs

On the ~maste~main:
```
$ ./python -m timeit -s 'from fractions import Fraction as Q; a=Q(11,5)' 'abs(a)'
50000 loops, best of 5: 4.98 usec per loop
```
With
```
diff --git a/Lib/fractions.py b/Lib/fractions.py
index 49a3f2841a..53086e7841 100644
--- a/Lib/fractions.py
+++ b/Lib/fractions.py
@@ -840,6 +840,8 @@ def __neg__(a):
 
     def __abs__(a):
         """abs(a)"""
+        if a._numerator >= 0:
+            return a
         return Fraction(abs(a._numerator), a._denominator, _normalize=False)
 
     def __int__(a, _index=operator.index):
```
I see:
```
$ ./python -m timeit -s 'from fractions import Fraction as Q; a=Q(11,5)' 'abs(a)'
1000000 loops, best of 5: 277 nsec per loop
```
On this pr:
```
$ ./python -m timeit -s 'from fractions import Fraction as Q; a=Q(11,5)' 'abs(a)'
100000 loops, best of 5: 2.71 usec per loop
```
So, there could be some speed bonus...  Ditto for `__neg__`.

For `__pos__` we, probably, could return self. Not sure which use case expects the type casting to the Fraction in this method...

File: Lib/fractions.py
Comment (eendebakpt) [reply Lib/fractions.py:820]: I understand. I find the original behaviour a bit strange, as `Fraction(5,1)/Fraction(0)` and `Fraction(0)**-2` also result in `ZeroDivisionError: Fraction(1, 0)`. Where did the `Fraction(1,0)` originate from? This is something internal, not something the user entered.

We could modify the error message a bit so that `Fraction(5,1)/Fraction(0)` results in 
```
`ZeroDivisionError: division by Fraction(0, 1)`
```
That is more in line with the behaviour for `int` or `float`. E.g. `2/0` results in 
```
ZeroDivisionError: division by zero
```

But if this PR just reproduces the original behaviour that is fine and we could create a separate issue for this.

Comment (skirpichev) [reply Lib/fractions.py:820]: > a bit strange

It says: "You are trying to create Fraction(foo, 0)".  Looks reasonable.

> Where did the Fraction(1,0) originate from?

I think, it's from 7736b5becd...

File: Lib/fractions.py
Comment (mdickinson) [reply Lib/fractions.py]: @eendebakpt 

> If Fraction is really immutable we could optimize here by returning self in case `a._numerator` is positive.

That would be a behaviour change for instances of `Fraction` subclasses: currently, `abs` on an instance of a strict `Fraction` subclass will return something of exact type `Fraction`.

**Comment by skirpichev** [comment]
Probably, from_float/decimal methods should be adapted to use this.  Decimal.as_integer_ratio is in lowest terms.  I guess, that same holds for the float method, but that's undocumented, unfortunately.

File: Lib/fractions.py
Comment (mdickinson) [reply Lib/fractions.py:820]: Given that this PR is supposed to be optimization-only, I agree with @skirpichev that the exception messages should stay unchanged. @eendebakpt If you feel strongly enough about the error message, let's open a separate issue for that.

File: Lib/fractions.py
Comment (mdickinson) [inline Lib/fractions.py]: Let's use `numerator` and `denominator` - there's no reason to introduce a naming inconsistency with the `__new__` method. It would also be good to have a docstring for the benefit of code readers - that docstring should explain that the purpose here is not to store an unnormalized fraction, but to bypass the normalization in situations where it's not needed. It might also be good to emphasize that it's an internal implementation detail that shouldn't be depended on by external users.

File: Lib/fractions.py
Comment (mdickinson) [inline Lib/fractions.py:186]: I'm wondering whether we'd need to keep the `_normalize` argument and issue a deprecation warning. It's true that it's undocumented, but I suspect that Hyrum's law applies here - it wouldn't be at all surprising if there are people passing `_normalize=False` for speed reasons.

**Comment by mdickinson** [comment]
@skirpichev 

> I guess, that same holds for the float method, but that's undocumented, unfortunately

It does indeed. It would be reasonable to document it, I think.

File: Lib/fractions.py
Comment (eendebakpt) [reply Lib/fractions.py:820]: I agree that the issues are better handled in a separate pr. Given the fact that this error message has been around for a very long time, i am not sure we should change it.

File: Lib/fractions.py
Comment (skirpichev) [reply Lib/fractions.py:186]: PEP 8 documents `_single_leading_underscore`-style names as "internal use" indicator.  No news (after 3c286e2e0d), no documentation...

So, I think it's not a part of the Fraction API and we can break it.

>  it wouldn't be at all surprising if there are people passing _normalize=False for speed reasons.

Indeed, [they do](https://github.com/search?l=Python&p=1&q=Fraction%28+_normalize%3DFalse&type=Code)  :(

Probably, there will be no harm to expose this kwarg as `normalize`, but I'm not sure if we should also issue a deprecation warning.  I'll revert this part for now.

Edit: IIUIC, almost everything (not equality or hashing) should work with or without normalization.

Comment (mdickinson) [reply Lib/fractions.py:186]: > PEP 8 documents [...]

I think you can and should assume that all core developers are very familiar with PEP 8. :-)

> So, I think it's not a part of the Fraction API and we can break it.

That's a rather dogmatic approach. On CPython, we should prefer a more pragmatic approach: if we have reasonable grounds to believe that a change might break user code then we should take reasonable steps to avoid that breakage. It's not friendly to knowingly break existing code and then tell people "well, it was your fault for using something undocumented".

**Review by eendebakpt — APPROVED** [review]

**Comment by skirpichev** [comment]
@mdickinson, `_normalize` part was reverted.  Does it make sense now?

**Comment by mdickinson** [comment]
> In this way we can end with supporting everything as part of the API.

Indeed. :-) As with most aspects of software development, careful judgement is required, evaluating the context and balancing the trade offs within that context. Few things are black and white. The fact that `_normalize` was intended to be private, and has a leading underscore, certainly means that removing it is easier than if it were public, but it's still something that shouldn't be done without thinking it through first. Hyrum's law is a real thing.

@skirpichev

> Does it make sense now?

Hmm, not really. :-) I don't want to leave a `_normalize` flags that's unused and untested, indefinitely. I think we have three options:

1. Leave this alone.
2. Introduce `_from_pair` (possibly after renaming - see below) and immediately remove the `_normalize` flag
3. Introduce `_from_pair` (ditto) and deprecate the `_normalize` flag.

Of those options, option 3 just seems like too much code churn for too little benefit.

There was a recent `discuss.python.org` thread in which the meaning of `_normalize` was misinterpreted, which pushes me a bit further towards option 2. Arguably, `_normalize` is too discoverable right now (since it shows up in IDE completions and tooltips), and there's a temptation to guess what it means. `_from_pair` would also be somewhat discoverable, but at least we get to add a docstring that clearly indicates that it's private, and the expected constraints for people who decide to use it anyway.

Screenshot 2023-02-26 at 12 52 41

File: Lib/fractions.py
Comment (mdickinson) [inline Lib/fractions.py]: Nitpick: I'd prefer `ints` to `int's`.

File: Lib/fractions.py
Comment (mdickinson) [inline Lib/fractions.py]: Can we rename `_from_pair` to something more descriptive? How about `_from_coprime_pair` or `_from_coprime_ints`?

**Comment by mdickinson** [comment]
Unless there are objections, let's go with option 2 and remove the `_normalize` flag.

**Review by mdickinson — CHANGES_REQUESTED** [review]
See comments for requested changes. Let's:

- remove `_normalize`
- rename `_from_pair` to something like `_from_coprime_pair`

**Comment by skirpichev** [comment]
> Arguably, _normalize is too discoverable right now (since it shows up in IDE completions and tooltips)

That is a more generic problem: we have PEP8's convention where \_single\_leading\_underscore is weak “internal use” indicator.  On another hand, these arguments are available for introspection and shown in the help() output, IDE tooltips and so on.

Probably, all above use the signature() function. Maybe it does make sense to add a `private=False` option for this function to hide private kwargs per default?

> There was a recent discuss.python.org thread in which the meaning of _normalize was misinterpreted

(Probably it was about [this thread](https://discuss.python.org/t/improve-support-for-unnormalized-fractions/24013).)  Indeed, there was an impression that all methods do support unnormalized fractions.  On another hand
1. there is no doubts that the argument was private
2. most methods actually do support this (arithmetics) and
3. much more can if we do a cheap normalization (ensure that denominator is positive)

So, I think the proposal might have some sense and it's possible to implement it with a little overhead for canonicalized fractions.  E.g. we can mark unnormalized fractions with a sticky flag (or flags) and do special steps to deal with this case in some methods.  Not sure it worth to have this in the stdlib: to really support unnormalized fractions almost every method should be rewritten (e.g. arithmetic methods do some normalization of the output, which we probably want to avoid for unnormalized fractions).

PS:
I have made the requested changes; please review again.

**Comment by eendebakpt** [comment]
@skirpichev @mdickinson The changes look good to me. With the removal of the (semi-private) argument `_normalize`, we could update the whatsnew entry with a note stating that it is removed.

File: Misc/NEWS.d/next/Library/2023-02-10-11-59-13.gh-issue-101773.J_kI7y.rst
Comment (eendebakpt) [inline Misc/NEWS.d/next/Library/2023-02-10-11-59-13.gh-issue-101773.J_kI7y.rst]:
```suggestion
Optimize :class:`fractions.Fraction` for small components. The private argument `_normalize` of the `Fraction` constructor has been removed.
```

Comment (skirpichev) [reply Misc/NEWS.d/next/Library/2023-02-10-11-59-13.gh-issue-101773.J_kI7y.rst]: Thanks.  Committed with a minor fix.

**Review by mdickinson — APPROVED** [review]
LGTM. I tweaked the docstring wording on `_from_coprime_ints` slightly.

I'll merge when CI completes.

**Comment by skirpichev** [comment]
Thanks to all reviewers.

**Pull request merged**
